tests/kernel/mutex/mutex_api/src/test_mutex_apis.c

```diff
@@ -140,18 +282,174 @@ void test_mutex_lock_unlock(void)
 	tmutex_test_lock_unlock(&kmutex);
 }
 
+/**
+ * @brief Test recursive mutex
+ * @details To verify that getting a lock of a mutex already locked will
+ * succeed and waiters will be unblocked only when the number of locks
+ * reaches zero.
+ * @ingroup kernel_mutex_tests
+ */
+void test_mutex_recursive(void)
+{
+	k_mutex_init(&mutex);
+
+	/**TESTPOINT: when mutex has no owner, we cannot unlock it */
+	zassert_true(k_mutex_unlock(&mutex) == -EINVAL,
+			"fail: mutex has no owner");
+
+	zassert_true(k_mutex_lock(&mutex, K_NO_WAIT) == 0,
+			"Failed to lock test_mutex");
+
+	/**TESTPOINT: The current thread does own the mutex */
+	zassert_true(k_mutex_lock(&mutex, K_NO_WAIT) == 0,
+			"Failed to recursively lock mutex");
+
+	k_mutex_unlock(&mutex);
+
+	/**TESTPOINT: when mutex has no owner, we cannot unlock it */
+	zassert_true(k_mutex_unlock(&mutex) == 0,
+			"Failed to recursively unlock mutex");
+}
+
+/**
+ * @brief Test mutex set amount of time the thread will wait
+ * @details To verify the mutex lock operation shall accept a timeout parameter
+ * indicating the maximum amount of time the thread will wait.
+ * @ingroup kernel_mutex_tests
+ */
+void test_mutex_param_timeout(void)
+{
+	k_mutex_init(&mutex);
+
+	/* we spawn a thread with holding mutex for 2000 ms */
+	k_thread_create(&tdata, tstack, STACK_SIZE,
+			(k_thread_entry_t)tThread_lock_with_timeout,
+			&mutex, &tdata, NULL,
+			K_PRIO_PREEMPT(0),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* wait for spawn thread to take action */
+	k_msleep(TIMEOUT);
+
+	/**TESTPOINT: The current thread does not own the mutex.*/
+	zassert_true(k_mutex_unlock(&mutex) == -EPERM,
+			"fail: current thread does not own the mutex");
+
+	/**TESTPOINT: test timeout with parameter K_NO_WAIT */
+	zassert_true(k_mutex_lock(&mutex, K_NO_WAIT) == -EBUSY,
+			"fail to lock K_NO_WAIT");
+
+	/**TESTPOINT: test timeout with parameter 500ms */
+	zassert_true(k_mutex_lock(&mutex, K_MSEC(TIMEOUT)) == -EAGAIN,
+			"fail to lock TIMEOUT");
```

Comment:
If you want to check a timeout for correctness, you need to measure the time taken.  Pure line coverage doesn't completely capture the behavior here.

Reply:
Oh wait, never mind.  You're looking for an -EAGAIN.  This is fine (but weird -- the timeout surely isn't inspected if the return is synchronous, right?  Is this really a code coverage gap?)

Reply:
Actually, this new test case "test_mutex_param_timeout", it can be totally covered by original "test_mutex_lock_unlock" one, except the check of  -EPERM, the test coverage only increase by this part. 
I add this new test case because I just try to emphasize the "timeout parameter" only. 
So, how do you think if I remove this new case and add the check of  -EPERM to other new case or existing test case?

Reply:
The test suite is large, the board set in CI keeps growing, and recently we've been adding a few different cycle-exact[1] emulator platforms.  I think reducing needless test creep is sort of a priority, yeah.

[1] That is: slow.  Things like renode, nsim and qemu -icount are all taking more CPU cycles than the older default qemu's.

Reply:
Ok, got it, I'll remove this new case and try to keep it clean, thanks!

Reply:
@andyross the latest master code seems have build problem in qemu_cortex_r5 platform today. I build it failed using pure master code.

Reply:
This built problem seems to be solved. I has updated the code.

Reply:
@andyross could you please help to review the modified code? I have fixed those points you mentioned.